src/main/java/com/chaos131/vision/LimelightCamera.java

```diff
@@ -0,0 +1,260 @@
+    /** NT Entry ID - Tag Distance Average */
+    private final int idxTagDistance = 9;
+    /** NT Entry ID - Tag Area */
+    private final int idxTagArea = 10;
+
+    @Override
+    public void recordMeasuredData() {
+        var data = m_botpose.getValue().getDoubleArray();
+        double timestampSeconds = Timer.getFPGATimestamp() - data[idxLatency] / 1000;
+        if (data == null || data[idxX] < EPSILON) {
+            m_mostRecentData = Optional.empty();
+            return;
+        }
+
+        var poseRotation = new Rotation3d(data[idxRoll]  * Math.PI / 180, 
+                                          data[idxPitch] * Math.PI / 180,
+                                          data[idxYaw]   * Math.PI / 180);
+
+        var visionPose = new Pose3d(data[idxX], data[idxY], data[idxZ], poseRotation);
+
+        if (m_offset != null) {
+            var cameraOffset = m_offset.get();
+            cameraOffset = cameraOffset.rotateBy(new Rotation3d(0, 0, data[idxYaw] * Math.PI / 180));
+            visionPose = new Pose3d(
+                new Translation3d(visionPose.getX() - cameraOffset.getX(),
+                                visionPose.getY() - cameraOffset.getY(),
+                                visionPose.getZ() - cameraOffset.getZ()),
+                new Rotation3d(0,//poseRotation.getX() - cameraOffset.getRotation().getX(),
+                                0,//poseRotation.getY() - cameraOffset.getRotation().getY(),
+                                poseRotation.getZ())//- cameraOffset.getRotation().getZ())
+            );
+        }
+
+        var deviation = calculateTranslationalDeviations(data[idxTagDistance], data[idxTagCount]);
+        var trackXYZ = MatBuilder.fill(Nat.N3(), Nat.N1(), new double[]{ deviation, deviation, 1 });
+
+        var conf = calculateConfidence(visionPose, (int)data[idxTagCount], data[idxTagDistance], deviation);
+        if (conf < m_specs.confidence_requirement) {
+            m_mostRecentData = Optional.empty();
+            return;
+        }
+
+        m_mostRecentData = Optional.of(new VisionData(visionPose, timestampSeconds, trackXYZ));
+        if (m_poseUpdator != null && m_useForOdometry) {
+            m_poseUpdator.accept(m_mostRecentData.get());
+        }
+    }
+
+    private double calculateTranslationalDeviations(double distance, double tagCount) {
+        var stddev = Math.pow(distance*m_specs.distance_scalar, m_specs.error_exponent);
+        // commented out for now until we know what's what
+        // stddev /= tagCount * VisionConstants.L3G.TagCountErrorScalar;
+        // stddev *= (1 + m_robotSpeedSupplier.get() * VisionConstants.LL3G.RobotSpeedErrorScalar);
+        return m_specs.error_multiplier * stddev + m_specs.minimum_error;
+    }
+
+    @Override
+    public void setOffsetHandler(Supplier<Pose3d> offsetHandler) {
+        // TODO Auto-generated method stub
+        throw new UnsupportedOperationException("Unimplemented method 'setOffsetHandler'");
```

Comment:
Should we do these in all the new functions that need completion?
(mostly for consistency)

Reply:
Yeah that's a good idea.